model.py

```diff
@@ -1,103 +1,1394 @@
+    # caclulate the location of the GP changepoints. Setting to initially
+    # unrealistically high values will ensure that the first time
+    # calcChangepoints is called, the changepoints are calculated.
+    _olddphi = 9e99
+    _oldq = 9e99
+    _oldrwd = 9e99
+
+    # _dist_cp is initially set to whatever, it will be overwritten anyway.
+    _dist_cp = 9e99
+
+    # Add the GP params
+    node_par_names = LCModel.node_par_names
+    node_par_names += ('ln_ampin_gp', 'ln_ampout_gp', 'ln_tau_gp')
+
+    def calcChangepoints(self, eclipse):
+        '''Caclulate the WD ingress and egresses, i.e. where we want to switch
+        on or off the extra GP amplitude.
+
+        Requires an eclipse object, since this is specific to a given phase
+        range.
+        '''
+
+        # Also get object for dphi, q and rwd as this is required to determine
+        # changepoints
+        dphi = self.ancestor_param_dict['dphi']
+        q = self.ancestor_param_dict['q']
+        rwd = self.ancestor_param_dict['rwd']
+
+        phi0 = eclipse.phi0.currVal
+
+        dphi_change = np.fabs(self._olddphi - dphi.currVal) / dphi.currVal
+        q_change = np.fabs(self._oldq - q.currVal) / q.currVal
+        rwd_change = np.fabs(self._oldrwd - rwd.currVal) / rwd.currVal
+
+        # Check to see if our model parameters have changed enough to
+        # significantly change the location of the changepoints.
+        if (dphi_change > 1.2) or (q_change > 1.2) or (rwd_change > 1.2):
+            # Calculate inclination
+            inc = roche.findi(q.currVal, dphi.currVal)
+
+            # Calculate wd contact phases 3 and 4
+            phi3, phi4 = roche.wdphases(q.currVal, inc, rwd.currVal, ntheta=10)
+
+            # Calculate length of wd egress
+            dpwd = phi4 - phi3
+
+            # Distance from changepoints to mideclipse
+            dist_cp = (dphi.currVal+dpwd)/2.
+
+            # save these values for speed
+            self._dist_cp = dist_cp
+            self._oldq = q.currVal
+            self._olddphi = dphi.currVal
+            self._oldrwd = rwd.currVal
+        else:
+            # Use the old values
+            dist_cp = self._dist_cp
+
+        # Find location of all changepoints
+        min_ecl = int(np.floor(eclipse.lc.x.min()))
+        max_ecl = int(np.ceil(eclipse.lc.x.max()))
+
+        eclipses = [e for e in range(min_ecl, max_ecl+1)
+                    if np.logical_and(e > eclipse.lc.x.min(),
+                                      e < 1+eclipse.lc.x.max()
+                                      )
+                    ]
+
+        changepoints = []
+        for e in eclipses:
+            # When did the last eclipse end?
+            egress = (e-1) + dist_cp + phi0
+            # When does this eclipse start?
+            ingress = e - dist_cp + phi0
+            changepoints.append([egress, ingress])
+
+        return changepoints
+
+    def create_GP(self, eclipse):
```

Comment:
Is there a reason why you haven't written a ```GPEclipse``` class instead of a GPLCModel? 

It seems more natural to me, and you don't have all the logic of passing around the eclipse to calculate the GP for, etc...

The ```GPEclipse``` would be the same as ```Eclipse``` but with a  different ```ln_like```

Reply:
See my comment above